### ISSUE TITLE
Jet Force Gemini Glide64 OpCode crash.

### ISSUE THREAD

**Issue by AmbientMalice** [project64/project64#346]
Trying to configure Glide64 while the game is running displays the following error.

![untitled](https://cloud.githubusercontent.com/assets/10704701/6912962/85be664e-d7bc-11e4-885f-349b9f37c46e.png)

**Comment by LegendOfDragoon** [comment]
Does that error happen with CPU interpreter? What RSP were you using and are you using HLE audio or LLE?

**Comment by LegendOfDragoon** [comment]
Just noticed your title says 2.2.0.2. You should use latest build. I believe it fixed the garbage text in error messages.

**Comment by AmbientMalice** [comment]
>Does that error happen with CPU interpreter?
Yes, but the memory address is different.

>What RSP were you using and are you using HLE audio or LLE?
1.7.0.13 LLE Jabo audio.

>Just noticed your title says 2.2.0.2. You should use latest build. I believe it fixed the garbage text in error messages.
Error is the same with 2.2.0.3\latest build.

The error only occurs when "Pause emulation when window not active" is disabled.

**Comment by LegendOfDragoon** [comment]
Alright, nice find. The problem seems to be related to Glide64. Doesn't seem to happen with any other video plugin.

**Comment by cxd4** [comment]
It's just another stupid C++ threading issue, caused by wxWidgets or whatever it's called.

You can stop the crashing by commenting out this line 1088 of Glide64 `Config.cpp`:
```c
mutexProcessDList->Lock();
```
To avoid other crashing that can randomly happen when closing config, also comment out 1156:
```c
mutexProcessDList->Unlock();
```
However I'm not convinced that I should send a pull request to do this, because it's really the plugin's fault for having such a volatile configuration system and wx thread hacking in the first place.  It's also quite possible that Gonetz had some reason in mind for freezing the ProcessDList thread calls while configuration is open, like allowing more DLL options to be configured without waiting until emulation shut-down and re-initialize to take effect or whatever.  I don't know what his reasons are or why do any of this C++ bullshit really.

So before committing a "fix" like this (commenting out the mutex lock and unlock) I think we need more detailed tests from people like Ambient to confirm whether this is safe to do for other games while the config is open.  I personally never like to have "Pause emulation when window is inactive" checked myself and find many benefits in leaving the emulator running while not looking at the window at front of my desktop.

**Comment by LegendOfDragoon** [comment]
Out of curiosity, does this happen with GlideN64? I cannot test that plugin myself, unfortunately :( .

**Comment by LegendOfDragoon** [comment]
Looks like it doesn't happen with GlideN64.

**Comment by oddMLan** [comment]
Cannot reproduce with current master.

>It's just another stupid C++ threading issue, caused by wxWidgets or whatever it's called.
68a8ed03696f3365430d8cca14d5084100165cd0

Most likely fixed by this commit.
WxWidgets is dead, long live WTL.

**Comment by LegendOfDragoon** [comment]
>Cannot reproduce with current master.

I think it was fixed by pausing the emulator.

**Comment by cxd4** [comment]
This bug reminds me of another "configure while running" crash that I had ended up fixing by tracing it to some inline asm functions that weren't using the correct callbacks.  Rewriting the asm to C fixed that.

May or may not be a separate issue from the fixed one but @AmbientMalice maybe you can close.

**Comment by LegendOfDragoon** [comment]
@AmbientMalice can you test this issue to see if it's fixed? I'm pretty sure the forced pausing prevents this crash.

**Comment by Frank-74** [comment]
I get crash in first intro with latest master and GLideN64.

Here's a savestate: https://www.dropbox.com/s/w8n49ty1ed5r7qu/Jet%20Force%20Gemini%20%28U%29.pj.zip?dl=1

A few seconds in, when the dog runs from the window, Project64 crashes with no log.

Disabling protect memory and enabling start changed, it gets further, but then crashes when the scene cuts to black.  Disabling advanced block linking then makes it past the fade to black, but it crashes again in the next scene with these errors:

2017/04/08 16:27:10.702 06368,Error,User Interface,CNotificationImp::DisplayError,Does sync effect Delay slot at 803012CC?
2017/04/08 16:27:31.301 06368,Error,User Interface,CNotificationImp::DisplayError,Unhandled R4300i opcode at: 803012D0
sync

Stopping Emulation !
2017/04/08 16:27:36.391 06368,Error,User Interface,CNotificationImp::DisplayError,Break point found at
N64System\Interpreter\InterpreterOps.cpp
2821
2017/04/08 16:27:41.784 06368,Error,User Interface,CNotificationImp::DisplayError,Break point found at
MemoryExceptionFilter.cpp
865
2017/04/08 16:27:43.708 06368,Error,Exception Handler,CMipsMemoryVM::FilterX86Exception,Invalid memory adderess: B46E0000
2017/04/08 16:27:43.777 06368,Error,User Interface,CNotificationImp::DisplayError,Break point found at
MemoryExceptionFilter.cpp
43
2017/04/08 16:27:45.315 06368,Error,User Interface,CNotificationImp::DisplayError,Unknown memory action

Emulation stopped

**Comment by Frank-74** [comment]
With only Cache enabled and advanced block linking disabled.  It gets another couple of scenes through the intro.  But then get this error:

2017/04/08 16:59:05.936 09852,Error,User Interface,CNotificationImp::DisplayError,Unhandled R4300i opcode at: BC416001
daddi	at, r0, 0x6001

Stopping Emulation !
2017/04/08 16:59:09.624 09852,Error,User Interface,CNotificationImp::DisplayError,Break point found at
N64System\Interpreter\InterpreterOps.cpp
2821

**Comment by LegendOfDragoon** [comment]
I think this issue should be closed since a workaround has already been implemented.

>Intro is fine with Interpreter. It plays to the end.

Sounds like the recompiler might need Protect Memory enabled and Protect Memory needs to be fixed. Then again, maybe GLideN64 is doing something to break exception handling.

**Comment by Frank-74** [comment]
The intro plays to the end with Protect Memory and Glide64 for PJ64.  But sections of the intro run at ~16VI's.
Intro crashes with Protect Memory and GLideN64.

There is no workaround for GLideN64.  Any changes made to get through the intro with a savestate breaks starting a new game.  I then get a WTF error starting a new game.

**Comment by LegendOfDragoon** [comment]
The issue @AmbientMalice is refering to, is when you open the GUI config while running the game. Both Glide64 and GLideN64 try to skip dlists while the GUI is open. The problem is that both plugins did an improper Dlist skipping implementation. They decided to just pause the emulator instead of skipping Dlists.

@Frank-74 Your issue has to do with Protect Memory and other possible cpu-recompiler issues. I tested with Glide64 and indeed got slowdowns in the Intro. Too much exception handling can cause serious performance issues. I personally would rather write per-game code to eliminate the need for using Protect Memory in various games. Although ironically, I have found that some games ( like Star Fox 64 (U) ) seem to have better performance with Protect Memory instead of Start Changed.
